creator-node/test/snapbackSM.test.js

```diff
@@ -810,4 +811,274 @@ describe('test SnapbackSM', function () {
+      return node === 'http://healthyNodeWithUserState.co' ? 1 : -1
+    }
+
+    const newReplicaSetNodes = await snapback.selectRandomReplicaSetNodes(
+      healthyReplicaSet,
+      numberOfUnhealthyReplicas,
+      healthyNodes,
+      wallet
+    )
+
+    // The missing nodes in healthyReplicaSet should've been replaced with 2 node from healthyNodes but not the node with user state
+    assert.strictEqual(newReplicaSetNodes.length, 2)
+    assert.ok(newReplicaSetNodes.every((newReplicaSetNode) => healthyNodes.includes(newReplicaSetNode)))
+    assert.ok(newReplicaSetNodes.every((newReplicaSetNode) => !healthyReplicaSet.has(newReplicaSetNode)))
+    assert.ok(newReplicaSetNodes.every((newReplicaSetNode) => newReplicaSetNode !== 'http://healthyNodeWithUserState.co'))
+  })
+
+  it('[selectRandomReplicaSetNodes] throw if insufficient number of new replica sets exist', async function () {
+    const { SnapbackSM: mockSnapback } = proxyquire(
+      '../src/snapbackSM/snapbackSM.js',
+      {
+        './StateMachineConstants': {
+          MAX_SELECT_NEW_REPLICA_SET_ATTEMPTS: 100
+        }
+      }
+    )
+    const snapback = new mockSnapback(nodeConfig, getLibsMock())
+
+    const healthyReplicaSet = new Set([
+      'http://healthyCn1.co'
+    ])
+    const numberOfUnhealthyReplicas = 2
+    const healthyNodes = [
+      'http://healthyCn1.co',
+      'http://healthyNodeWithUserState.co',
+      'http://healthyCn4.co',
+    ]
+    const wallet = '0x00fc5bff87afb1f15a02e82c3f671cf5c9ad9e6d'
+
+    // Mock the user having no state on any of the healthy nodes
+    snapback._retrieveClockValueForUserFromReplica = async (node, _) => {
+      return node === 'http://healthyNodeWithUserState.co' ? 1 : -1
+    }
+
+    // Only 1 node is healthy and doesn't have user state, so it should throw when it can't find a second node
+    await assert.rejects(async () => {
+      await snapback.selectRandomReplicaSetNodes(
+        healthyReplicaSet,
+        numberOfUnhealthyReplicas,
+        healthyNodes,
+        wallet
+      )
+    }, /.*Not enough healthy nodes found to issue new replica set.*100 attempts.*/
+    ).catch((reason) => {
+      assert.ok(false, reason)
+    })
+  })
+
+  it('[retrieveClockStatusesForUsersAcrossReplicaSet] returns expected clock values and updates unhealthyPeers', async function () {
+    nodeConfig.set('maxBatchClockStatusBatchSize', 2)
+
+    const expectedReplicaToClockValueMap = {
+      'http://healthyCn1.co': {
+        'wallet1': 1,
+        'wallet2': 2,
+        'wallet3': 3,
+        'wallet4': 4,
+        'wallet5': 5
+      },
+      'http://healthyCn2.co': {
+        'wallet1': 10,
+        'wallet2': 20
+      }
+    }
+
+    const { SnapbackSM: mockSnapback } = proxyquire(
+      '../src/snapbackSM/snapbackSM.js',
+      {
+        'axios': async (params) => {
+          const { baseURL: replica, data } = params
+          if (replica === 'http://unhealthyCn.co') {
+            return { data: {} }
```

Comment:
should this throw an error to represent unhealthy?

Reply:
it shouldn't matter much. since it doesn't have `data.data.users` the function throws an error anyway. this error caught and triggers the replica to be added to `unhealthyPeers`, which we assert against